test/jdk/java/lang/String/StringRacyConstructor.java

```diff
@@ -0,0 +1,418 @@
+        // until we hit the race condition
+        int i = 0;
+        while (true) {
+            i++;
+            String s = new String(codePoints, 0, len);
+            if ((s.charAt(0) < 256 && !original.equals(s)) || i > 1_000_000) {
+                thread.interrupt();
+                try {
+                    thread.join();
+                } catch (InterruptedException ie) {
+                    // ignore interrupt
+                }
+                if (i >= 1_000_000) {
+                    System.err.printf("Unable to produce a UTF16 string in %d iterations: %s%n", i, original);
+                }
+                return s;
+            }
+        }
+    }
+
+    /**
+     * Returns a string created from a codepoint array that has been racily
+     * modified to contain high and low surrogates. The string is a different length
+     * than the original due to the surrogate encoding.
+     */
+    public static String racyStringConstructionCodepointsSurrogates(String original) throws ConcurrentModificationException {
+        if (original.chars().max().orElseThrow() > 256) {
+            throw new IllegalArgumentException(
+                    "Can only work with latin-1 Strings");
+        }
+
+        int len = original.length();
+        int[] codePoints = new int[len];
+        for (int i = 0; i < len; i++) {
+            codePoints[i] = original.charAt(i);
+        }
+
+        // In another thread, flip the first character back
+        // and forth between being latin-1 or as a surrogate pair.
+        Thread thread = new Thread(() -> {
+            while (!Thread.interrupted()) {
+                codePoints[0] ^= 0x10000;
+            }
+        });
+        thread.start();
+
+        // at the same time call the String constructor,
+        // until we hit the race condition
+        int i = 0;
+        while (true) {
+            i++;
+            String s = new String(codePoints, 0, len);
+            if ((s.length() != original.length()) || i > 1_000_000) {
+                thread.interrupt();
+                try {
+                    thread.join();
+                } catch (InterruptedException ie) {
+                    // ignore interrupt
+                }
+                if (i >= 1_000_000) {
+                    System.err.printf("Unable to create a string in %d iterations: %s%n", i, original);
+                }
+                return s;
+            }
+        }
+    }
+
+    // A CharSequence that returns characters from a string and throws IllegalArgumentException
+    // when the character requested is 0xFFFD (the replacement character)
+    // The string contents determine when the exception is thrown.
+    static class ThrowingCharSequence implements CharSequence {
+        private final String aString;
+
+        ThrowingCharSequence(String aString) {
+            this.aString = aString;
+        }
+
+        @Override
+        public int length() {
+            return aString.length() + 1;
```

Comment:
Not sure why ` + 1`.

Reply:
Removed; it was a leftover from a prior way to throw an exception during `CharSequence.charAt(n)`.